dns_sd_windows.c

```diff
@@ -202,14 +204,17 @@ int dnssd_find_hosts(struct dns_sd_discovery_data **ddata)
 
 	IIO_DEBUG("DNS SD: Start service discovery.\n");
 
-	*ddata = zalloc(sizeof(**ddata));
-	if (!*ddata)
+	d = zalloc(sizeof(*d));
+	if (!d)
 		goto out_wsa_cleanup;
```

Comment:
Since you added a `d` variable instead of using `*ddata` you are now leaking memory, because in all error cases the dns_sd_discovery_data won't be freed by the parent function.

Reply:
d is passed back as a pointer, like all other implementations (avahi, and bonjour), so when the parent frees the variable it's fine.

Reply:
Memory is allocated inside each implementation (`dns_sd_avahi.c` `dns_sd_bonjour.c` `dns_sd_windows.c`) in the `dnssd_find_hosts()` function. It's cleared when the top level calls `dnssd_free_all_discovery_data()`

You are right - I was missing freeing it in the error paths. That is fixed now.

Reply:
Not entirely fixed, the `d` is not freed if iio_mutex_create() fails.
Also - why not simply set `*ddata = d;` right after `d` is allocated? Then you wouldn't have to call `dnssd_free_all_discovery_data` everywhere.

Reply:
Yeah, you have sharper eyes/better ideas that I do...

done.